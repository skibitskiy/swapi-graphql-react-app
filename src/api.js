const baseUrl = 'https://swapi.dev/api/';

const fetchData = async (type, page = 1) => {
    const url = new URL(type, baseUrl);
    url.searchParams.append('page', page);
    const response = await fetch(url).then((res) => res.json());
    return response
}

const fetchResources = async (type, page = 1) => {
    const { results: resources } = await fetchData(type, page);
    return resources
}

const fetchTitles = async (type, page) => {
    const resources = await fetchResources(type, page);
    return resources.map((resource) => {
        const { name, title, url } = resource;
        return {
            name: name || title,
            url
        }
    });
}

const fetchPageCount = async (type) => {
    const { count } = await fetchData(type);
    return Math.ceil((parseInt(count) || 0) / 10);
}

const fetchResource = async (url) => {
    const response = await fetch(url).then((res) => res.json());
    return response
}

export {
    fetchData,
    fetchResources,
    fetchTitles,
    fetchPageCount,
    fetchResource
}
