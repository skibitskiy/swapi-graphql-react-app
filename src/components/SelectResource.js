import React, { memo } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import { Select, InputLabel, MenuItem, Paper } from '@material-ui/core';

const useStyles = makeStyles(() => ({
  paper: {
    padding: '16px'
  }
}));

function SelectResource(props) {
  const styles = useStyles();
  const { resource, onResourceChange } = props;

  return (
    <Paper className={styles.paper}>
      <InputLabel>Выбери объект</InputLabel>
      <Select value={resource} onChange={onResourceChange}>
        <MenuItem value='films'>Фильмы</MenuItem>
        <MenuItem value='people'>Персонажи</MenuItem>
        <MenuItem value='planets'>Планеты</MenuItem>
        <MenuItem value='species'>Расы</MenuItem>
        <MenuItem value='starships'>Звездолеты</MenuItem>
        <MenuItem value='vehicles'>Транспорт</MenuItem>
      </Select>
    </Paper>  
  );
}

export default memo(SelectResource);
