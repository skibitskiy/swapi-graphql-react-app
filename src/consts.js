const FILMS = 'films';
const PEOPLE = 'people';
const PLANETS = 'planets';
const SPECIES = 'species';
const STARSHIPS = 'starships';
const VEHICLES = 'vehicles';
const CHOOSE_OBJECT = 'Выбери объект';

const selectItems = {
    [FILMS]: 'Фильмы',
    [PEOPLE]: 'Персонажи',
    [PLANETS]: 'Планеты',
    [SPECIES]: 'Расы',
    [STARSHIPS]: 'Звездолеты',
    [VEHICLES]: 'Транспорт'
};

export {
    selectItems,
    FILMS,
    PEOPLE,
    PLANETS,
    SPECIES,
    STARSHIPS,
    VEHICLES,
    CHOOSE_OBJECT
}