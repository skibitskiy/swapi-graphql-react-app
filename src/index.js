import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
import App from './components/App.js';
import * as serviceWorker from './serviceWorker';
import { ApolloClient, InMemoryCache, ApolloProvider  } from '@apollo/client';

const serverURI = process.env.SERVER;
const client = new ApolloClient({
  uri: serverURI || 'http://localhost:1166/graphql',
  cache: new InMemoryCache({
    possibleTypes: {
      IResource: ['Film', 'People', 'Species', 'Vehicle', 'Starship', 'Planet']
    }
  })
});

ReactDOM.render(
  <ApolloProvider client={client}>
    <React.StrictMode>
      <App />
    </React.StrictMode>
  </ApolloProvider>,
  document.getElementById('root')
);

// If you want your app to work offline and load faster, you can change
// unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: https://bit.ly/CRA-PWA
serviceWorker.unregister();
