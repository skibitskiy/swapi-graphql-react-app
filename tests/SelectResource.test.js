import '@testing-library/jest-dom/extend-expect';
import { within } from '@testing-library/dom';
import React from 'react';
import * as test from '@testing-library/react';
import user from '@testing-library/user-event';

import SelectResource from '../src/components/SelectResource';
import { selectItems, PEOPLE, FILMS } from '../src/consts';

describe('SelectResource component', () => {
    it('Snapshot', async () => {
        const props = {
            resource: 'people',
            onResourceChange: jest.fn()
        };        
        const { container } = test.render(<SelectResource {...props}/>);
        expect(container).toMatchSnapshot();
    });

    it('OnClick check', () => {
        const props = {
            resource: 'people',
            onResourceChange: jest.fn()
        };        
        const { baseElement } = test.render(<SelectResource {...props}/>);

        // click to expand select
        test.act(() => {
            const selectElement = test.screen.getByText(selectItems[PEOPLE], { exact: false });
            user.click(selectElement);
        });

        // expect that select expanded (all options exists in the document)
        Object.values(selectItems).forEach((resourceType) => {
            const optionsElement = baseElement.children[1];
            expect(within(optionsElement).getByText(resourceType)).toBeInTheDocument();
        });

        // select new value
        test.act(() => {
            const optionElement = test.screen.getByText(selectItems[FILMS]);
            user.click(optionElement);
        });
        expect(props.onResourceChange.mock.calls.length).toBe(1);
    });
})
