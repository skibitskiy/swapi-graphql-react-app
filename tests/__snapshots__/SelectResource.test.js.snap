// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectResource component Snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root makeStyles-paper-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated"
    >
      Выбери объект
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline"
    >
      <div
        aria-haspopup="listbox"
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        role="button"
        tabindex="0"
      >
        Персонажи
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput"
        tabindex="-1"
        value="people"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;
